pkg/forwarders/tlsforwarder_test.go

```diff
@@ -75,126 +138,70 @@ f3cb9gYaLWdmvkx8p3g=
-
 	tlsCert, err := tls.X509KeyPair([]byte(rootPEM), []byte(keyPEM))
 	require.NoError(t, err)
 
-	listener, err := tls.Listen("tcp4", "localhost:0",
+	listener, err := tls.Listen(
+		"tcp4",
+		addr,
 		&tls.Config{
-			RootCAs:      roots,
-			Certificates: []tls.Certificate{tlsCert},
-			// Tried to actually verify the certificates in this test but was getting the following error
-			// event though 127.0.0.1 and localhost were added to alt_names in openssl config:
-			// failed to connect to reporting server: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
+			Certificates:       []tls.Certificate{tlsCert},
 			InsecureSkipVerify: true, //nolint:gosec
 		},
 	)
 	require.NoError(t, err)
-	require.NotNil(t, listener)
 	t.Logf("TLS server address %s", listener.Addr().String())
 
-	var wg sync.WaitGroup
-	go acceptLoop(t, listener, &wg,
-		[]string{
-			"<14>signal=test-signal;key=value;\n",
-			"<14>signal=test-signal-2;key=value;\n",
-		},
-	)
-
-	serializer := serializers.NewSemicolonDelimited()
-	require.NotNil(t, serializer)
-
-	log := logrusr.New(logrus.New())
-	tf, err := NewTLSForwarder(listener.Addr().String(), log,
-		func(c *tls.Config) {
-			c.RootCAs = roots
-			c.Certificates = []tls.Certificate{tlsCert}
-			c.InsecureSkipVerify = true
-		},
-	)
-	require.NoError(t, err)
-	require.NotNil(t, tf)
-
-	consumer := telemetry.NewConsumer(serializer, tf)
-
-	m, err := telemetry.NewManager(
-		"test-ping",
-		telemetry.OptManagerPeriod(time.Hour),
-		telemetry.OptManagerLogger(log),
-	)
-	require.NoError(t, err)
-
-	w := telemetry.NewWorkflow("test1")
-	p, err := provider.NewFixedValueProvider("test1-provider", types.ProviderReport{
-		"key": "value",
-	})
-	require.NoError(t, err)
-	w.AddProvider(p)
-	m.AddWorkflow(w)
-	require.NoError(t, m.AddConsumer(consumer))
-	require.NoError(t, m.Start())
-
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal"))
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal-2"))
-	wg.Wait()
+	return telemetryServer{
+		listener: listener,
+	}
+}
 
-	require.NoError(t, listener.Close())
-	m.Stop()
+func (ts telemetryServer) Addr() string {
+	return ts.listener.Addr().String()
 }
 
-func acceptLoop(t *testing.T, l net.Listener, wg *sync.WaitGroup, expectedData []string) {
+func (ts telemetryServer) RunAndAssertExpectedData(ctx context.Context, t *testing.T, wg *sync.WaitGroup, expectedData []string) {
 	t.Log("server: accepting...")
-
-	// Accept just once because TLSForwarder persists the connection
-	conn, err := l.Accept()
-	if err != nil {
-		t.Logf("server: Accept() returned error: %v", err)
-		return
+	receivedData := make(chan string, len(expectedData))
+	wg.Add(1)
+	go func() {
+		for {
+			conn, err := ts.listener.Accept()
+			if err != nil {
+				t.Logf("server: Accept() returned error: %v", err)
+				return
+			}
+			go handleConnection(t, conn, receivedData)
+		}
+	}()
+	defer ts.listener.Close()
```

Comment:
food for thought: this could be closed in `newTelemetryTestServer()` using provided `*testing.T`'s `.Cleanup()`

Reply:
yes, but in the next step when I'm going to test what happen when connection is lost I'm going to create second instance of this server, so I want to clean as soon as possible

Reply:
In case of test that pass for both part before and after losing connection, it will read all messages successfully. But yeah it's nit